sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -2656,6 +2656,14 @@ object SQLConf {
       .checkValue(_ > 0, "The difference must be positive.")
       .createWithDefault(4)
 
+  val ALLOW_MISSING_COLUMNS_IN_UNION_BY_NAME =
+    buildConf("spark.sql.allowMissingColumnsInUnionByName")
+    .doc("If this config is enabled, `Dataset.unionByName` allows different set of column names " +
+      "between two Datasets. Missing columns at each side, will be filled with null values.")
```

Comment:
It seems not a breaking change if this case fails before and now we allow it by filling missing columns with nulls. Do we really need a config? cc @gatorsmile @bart-samwel

Reply:
I think there's value in forbidding this behavior by default, because that might hide logical errors (unioning two totally different schemas). But this isn't a matter of old/new, and it shouldn't be decided by global configs -- users would need to set the old or the new behavior *on a case by case basis*. Why not add a boolean parameter to `unionByName` to allow/disallow, defaulting to false (disallow)? Or maybe even add this as a separate function `unionByNameAllowMissing`? (Boolean arguments are not terribly informative to readers, so I'd actually lean towards having a separate name for this.)

Reply:
Adding a boolean parameter sounds better than new method `unionByNameAllowMissing`. But I have concern about calling with default parameter from Java.

Reply:
Seems like `Dataset` already has many APIs taking a boolean parameter. I'm OK with adding a `allowMissingColumns` parameter to `union`.

Reply:
But a default value parameter seems bad for Java caller?

Reply:
We can add an overload method instead of using default parameter value.

Reply:
+1 for overloading.